tokio/src/sync/rwlock.rs

```diff
@@ -0,0 +1,151 @@
+//! An asynchronous `RwLock`-like type.
+//!
+//! This module provides [`RwLock`], a type that acts similarly to an asynchronous `RwLock`.
+//!
+//! This allows you to do something along the lines of:
+//!
+//! ```rust,no_run
+//! use tokio::sync::RwLock;
+//! use std::sync::Arc;
+//!
+//! #[tokio::main]
+//! async fn main() {
+//!     let data1 = Arc::new(RwLock::new(0));
+//!     let data2 = data1.clone();
+//!
+//!     tokio::spawn(async move {
+//!         let mut lock = data2.write().await;
+//!         *lock += 1;
+//!     });
+//!
+//!     let mut lock = data1.write().await;
+//!     *lock += 1;
+//! }
+//! ```
+//!
+//! [`RwLock`]: struct.RwLock.html
+//! [`RwLockReadGuard`]: struct.RwLockReadGuard.html
+//! [`RwLockWriteGuard`]: struct.RwLockWriteGuard.html
+
+use crate::sync::semaphore;
+use crate::sync::Mutex;
+use futures_util::future::poll_fn;
+use std::cell::UnsafeCell;
+use std::ops;
+
+const MAX_READS: usize = 32;
+/// An asynchronous readers writer lock usefull for protecting shared data
+///
+/// Each lock has a type parameter (`T`) which represents the data that it is protecting. The data
+/// can only be accessed through the RAII guards returned from `read` and `write`, which
+/// guarantees that the data is only ever modified when the rwlock is with write lock.
+#[derive(Debug)]
+pub struct RwLock<T> {
+    //semaphore to coordinate read and write access to T
+    s: semaphore::Semaphore,
+
+    //inner data T
+    c: UnsafeCell<T>,
+
+    //mutex to coordinate writes
+    w: Mutex<()>,
+}
+
+// As long as T: Send, it's fine to send and share Mutex<T> between threads.
+// If T was not Send, sending and sharing a Mutex<T> would be bad, since you can access T through
+// Mutex<T>.
+unsafe impl<T> Send for RwLock<T> where T: Send {}
+unsafe impl<T> Sync for RwLock<T> where T: Send {}
+unsafe impl<'a, T> Sync for RwLockReadGuard<'a, T> where T: Send + Sync {}
+unsafe impl<'a, T> Sync for RwLockWriteGuard<'a, T> where T: Send + Sync {}
+
+/// A handle to a shared access to `Rwlock`.
```

Comment:
More natural wording would be "A shared-access handle to a locked `RwLock`.

Reply:
Re: all the various doc comments...since this API is modeled after `std::sync::RwLock`, why don't we just reuse the same doc comments (and modify them where appropriate, e.g., adding `await` to the examples and so on).